cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java

```diff
@@ -170,22 +219,96 @@ private void ensureTopicsCreated(Map<String, ?> config) {
                               numberOfBrokersInCluster) : _sampleStoreTopicReplicationFactor;
 
       ensureTopicCreated(zkUtils, topics.keySet(), _partitionMetricSampleStoreTopic, partitionSampleRetentionMs,
-                         replicationFactor);
+                         replicationFactor, _partitionSampleStoreTopicPartitionCount);
       ensureTopicCreated(zkUtils, topics.keySet(), _brokerMetricSampleStoreTopic, brokerSampleRetentionMs,
-                         replicationFactor);
+                         replicationFactor, _brokerSampleStoreTopicPartitionCount);
     } finally {
       KafkaCruiseControlUtils.closeZkUtilsWithTimeout(zkUtils, 10000);
     }
   }
 
-  private void ensureTopicCreated(ZkUtils zkUtils, Set<String> allTopics, String topic, long retentionMs, int replicationFactor) {
+  /**
+   * Increase the replication factor of a Kafka topic through adding new replicas in a rack-aware, round-robin way.
+   * If Zookeeper does not have rack information about brokers, then rack-aware property is not guaranteed.
+   * In this case it is only guaranteed that new replicas are added to brokers which do not currently host the partition.
+   * @param zkUtils ZkUtils class to use to increase replication factor.
+   * @param topicMetadata Topic metadata stored in Zookeeper.
+   * @param replicationFactor The replication factor to set for the topic.
+   * @param topic The topic to apply the change.
+   * @param props The properties to set for the topic.
+   */
+  private void increaseTopicReplicationFactor(ZkUtils zkUtils,
+                                              MetadataResponse.TopicMetadata topicMetadata,
+                                              int replicationFactor,
+                                              String topic,
+                                              Properties props) {
+    Map<String, List<Integer>> brokersByRack = new HashMap<>();
+    Map<Integer, String> rackByBroker = new HashMap<>();
+    for (BrokerMetadata bm:
+         JavaConversions.seqAsJavaList(AdminUtils.getBrokerMetadatas(zkUtils, RackAwareMode.Enforced$.MODULE$, Option.empty()))) {
+      // If the rack is not specified, we use the broker id info as rack info.
+      String rack = bm.rack().isEmpty() ? String.valueOf(bm.id()) : bm.rack().get();
+      brokersByRack.putIfAbsent(rack, new ArrayList<>());
+      brokersByRack.get(rack).add(bm.id());
+      rackByBroker.put(bm.id(), rack);
+    }
+    if (replicationFactor > brokersByRack.size()) {
+      throw new RuntimeException("Unable to increase topic " + topic + " replica factor to " + replicationFactor
```

Comment:
Please refer to my previous comment regarding failing to start due to exceptions for changing configs of the created topics.